core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceRedundantJoinWithSource.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableSet;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.JoinNode;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.sql.planner.iterative.rule.Util.restrictOutputs;
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isAtMost;
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isScalar;
+import static io.trino.sql.planner.plan.Patterns.join;
+
+/**
+ * This rule transforms plans with joins, where:
+ * - one of the sources is scalar and produces no output symbols.
+ * In case of LEFT or RIGHT join, it has to be the inner source.
+ * In case of FULL or INNER join, it can be either source.
+ * - the other join source is at least scalar (not known to be empty).
+ * <p>
+ * The join is replaced with the other source and an optional
+ * pruning projection.
+ */
+public class ReplaceRedundantJoinWithSource
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join();
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        boolean leftSourceScalarWithNoOutputs = node.getLeft().getOutputSymbols().isEmpty() && isScalar(node.getLeft(), context.getLookup());
+        boolean rightSourceScalarWithNoOutputs = node.getRight().getOutputSymbols().isEmpty() && isScalar(node.getRight(), context.getLookup());
+
+        boolean leftSourceNotKnownEmpty = !isAtMost(node.getLeft(), context.getLookup(), 0);
+        boolean rightSourceNotKnownEmpty = !isAtMost(node.getRight(), context.getLookup(), 0);
+
+        switch (node.getType()) {
+            case INNER:
+                if (node.isCrossJoin()) {
+                    if (leftSourceScalarWithNoOutputs && rightSourceNotKnownEmpty) {
```

Comment:
Why does it matter whether the right side produces rows? We should be able to substitute with that source unconditionally.

Reply:
If the right source is _known empty_, we have another rule to take care of it: `RemoveRedundantJoin`. That rule folds the whole join into empty Values. I added this condition to avoid overlapping rules, especially that their immediate results may be different.

Reply:
I don't think we should complicate this rule just to keep it from overlapping with the other one. The end result will be the same if the subquery produces no output.

Maybe we can re-organize both rules to cut across a different axis, but I haven't thought deeply about it.

Reply:
I thought about it and I can see no better way to split the optimization into rules.

To give you a full picture, there's one more rule of this kind I'd like to add. It will handle the cases where one join source is a single constant row. It could be appended as a projection instead of join.

That rule will cause the same overlapping issue as this one.

My proposition is:
- do not let the rules overlap
- separate the overlap check from the main rewrite logic. It should fold into a short simple "if". Add a comment explaining why certain cases are excluded.

What do you think?

Reply:
Ok, I'm fine with that as long as the checks are not all mixed into the rewrite logic.